### ISSUE TITLE
Crash wss and topology hiding

### ISSUE THREAD

**Issue by hydrosine** [OpenSIPS/opensips#1095]
Found another crash,
Seems to happen on the 200OK which is sent by the Websockets client.
Can replicate it every time.
This is a call initiated towards the WSS client. 
The other way around, Client -> opensips works.

I see some other tickets related towards TCP sessions could be a duplicate!

latest commit for this version 2f688b5fa48a4b84e0d5efd9367a0e50c8bf3c2c

```
Apr  7 14:32:07 srv-ams5-cod-wss-01 /usr/sbin/opensips[29106]: [INVITE] [100] in onreply NAT
Apr  7 14:32:07 srv-ams5-cod-wss-01 /usr/sbin/opensips[29106]: [INVITE] [180] in onreply NAT
Apr  7 14:32:07 srv-ams5-cod-wss-01 kernel: [259505.655202] opensips[29106]: segfault at 28 ip 00007f6775af23a7 sp 00007fff8a620380 error 6 in topology_hiding.so[7f6775af0000+11000]
Apr  7 14:32:07 srv-ams5-cod-wss-01 /usr/sbin/opensips[29115]: CRITICAL:core:receive_fd: EOF on 51
Apr  7 14:32:07 srv-ams5-cod-wss-01 /usr/sbin/opensips[29092]: INFO:core:handle_sigs: child process 29106 exited by a signal 11
```
Backtrace:
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f6775af23a7 in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/topology_hiding.so
(gdb) bt full
#0  0x00007f6775af23a7 in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/topology_hiding.so
No symbol table info available.
#1  0x00007f6775af2da5 in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/topology_hiding.so
No symbol table info available.
#2  0x00007f6775afaa8e in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/topology_hiding.so
No symbol table info available.
#3  0x00007f6775afb2ae in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/topology_hiding.so
No symbol table info available.
#4  0x00007f6775d1319b in run_dlg_callbacks () from /usr/lib/x86_64-linux-gnu/opensips/modules/dialog.so
No symbol table info available.
#5  0x00007f6775d261fb in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/dialog.so
No symbol table info available.
#6  0x00007f6777f03e21 in run_trans_callbacks () from /usr/lib/x86_64-linux-gnu/opensips/modules/tm.so
No symbol table info available.
#7  0x00007f6777f11a3a in relay_reply () from /usr/lib/x86_64-linux-gnu/opensips/modules/tm.so
No symbol table info available.
#8  0x00007f6777f15545 in reply_received () from /usr/lib/x86_64-linux-gnu/opensips/modules/tm.so
No symbol table info available.
#9  0x000000000043b60c in forward_reply ()
No symbol table info available.
#10 0x00000000004720b2 in receive_msg ()
No symbol table info available.
#11 0x00007f67750628d0 in ?? () from /usr/lib/x86_64-linux-gnu/opensips/modules/proto_wss.so
No symbol table info available.
#12 0x00000000005392c9 in ?? ()
No symbol table info available.
#13 0x000000000053c7df in tcp_worker_proc_loop ()
No symbol table info available.
#14 0x0000000000534966 in tcp_start_processes ()
No symbol table info available.
#15 0x000000000041c3f3 in main ()
No symbol table info available.
```

**Comment by bogdan-iancu** [comment]
Hi @hydrosine , please indicate the OpenSIPS version and revision number.
Also, can you install the debug symbols so you get a better backtrace ?
I see the TH module is used - if you disabled it, is it still crashing ?

**Comment by hydrosine** [comment]
@bogdan-iancu commit is attached to first post, it's one of the latest 2.3.0.
same crash with debug enabled:

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f983fe0b3a7 in delete_existing_contact (msg=msg@entry=0x7f98448ff238) at topo_hiding_logic.c:326
326     topo_hiding_logic.c: No such file or directory.
(gdb) bt full
#0  0x00007f983fe0b3a7 in delete_existing_contact (msg=msg@entry=0x7f98448ff238) at topo_hiding_logic.c:326
        offset = 294
        len = <optimized out>
        lump = 0x7f9844900b68
        crt = 0x0
        prev_crt = 0x0
        a = <optimized out>
        foo = <optimized out>
        __FUNCTION__ = "delete_existing_contact"
#1  0x00007f983fe0bda5 in topo_dlg_replace_contact (msg=msg@entry=0x7f98448ff238, dlg=dlg@entry=0x7f9842c81760) at topo_hiding_logic.c:565
        prefix = 0x7f9844903998 "<sip:\021"
        suffix = 0x7f9844900d40 ";did=12c.988dd8c2>\360\360\360\360\360\360\071\001"
        p = 0x7f9844900d52 "\360\360\360\360\360\360\071\001"
        p_init = 0x7f9844900d40 ";did=12c.988dd8c2>\360\360\360\360\360\360\071\001"
        ct_username = 0x0
        prefix_len = 5
        suffix_len = 18
        ct_username_len = <optimized out>
        n = <optimized out>
        i = <optimized out>
        ctu = {user = {s = 0x0, len = 0}, passwd = {s = 0x0, len = 0}, host = {s = 0x0, len = 0}, port = {s = 0x0, len = 0}, params = {s = 0x0, len = 0}, headers = {s = 0x0, len = 0}, port_no = 0, proto = 0, 
          type = ERROR_URI_T, transport = {s = 0x0, len = 0}, ttl = {s = 0x0, len = 0}, user_param = {s = 0x0, len = 0}, maddr = {s = 0x0, len = 0}, method = {s = 0x0, len = 0}, lr = {s = 0x0, len = 0}, r2 = {
            s = 0x0, len = 0}, gr = {s = 0x0, len = 0}, transport_val = {s = 0x0, len = 0}, ttl_val = {s = 0x0, len = 0}, user_param_val = {s = 0x0, len = 0}, maddr_val = {s = 0x0, len = 0}, method_val = {s = 0x0, 
            len = 0}, lr_val = {s = 0x0, len = 0}, r2_val = {s = 0x0, len = 0}, gr_val = {s = 0x0, len = 0}, u_name = {{s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, 
              len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}}, u_val = {{s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, 
            {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}, {s = 0x0, len = 0}}, u_params_no = 0}
        contact = <optimized out>
        el = <optimized out>
        it = <optimized out>
        rr_param = <optimized out>
        lump = <optimized out>
        __FUNCTION__ = "topo_dlg_replace_contact"
#2  0x00007f983fe13a8e in dlg_th_onreply (dlg=dlg@entry=0x7f9842c81760, rpl=0x7f98448ff238, req=0x7f9842c6c8d8, init_req=init_req@entry=1, dir=dir@entry=2) at topo_hiding_logic.c:1116
        lmp = <optimized out>
        size = <optimized out>
        route = <optimized out>
        leg = <optimized out>
        __FUNCTION__ = "dlg_th_onreply"
#3  0x00007f983fe142ae in topo_dlg_initial_reply (dlg=0x7f9842c81760, type=<optimized out>, params=0x7f984028eca0 <params>) at topo_hiding_logic.c:1021
        t = <optimized out>
#4  0x00007f984002c19b in run_dlg_callbacks (type=type@entry=256, dlg=dlg@entry=0x7f9842c81760, msg=msg@entry=0x7f98448ff238, dir=dir@entry=2, dlg_data=dlg_data@entry=0x0, locked=locked@entry=0) at dlg_cb.c:275
        cb = 0x7f9842c81a58
---Type <return> to continue, or q <return> to quit---
        __FUNCTION__ = "run_dlg_callbacks"
#5  0x00007f984003f1fb in dlg_onreply (t=<optimized out>, type=<optimized out>, param=<optimized out>) at dlg_handlers.c:461
        rpl = 0x7f98448ff238
        req = <optimized out>
        dlg = 0x7f9842c81760
        new_state = 1062747691
        old_state = 1150288352
        unref = 1062747965
        event = <optimized out>
        mangled_from = {s = 0x0, len = 0}
        mangled_to = {s = 0x0, len = 0}
        req_out_buff = <optimized out>
        __FUNCTION__ = "dlg_onreply"
#6  0x00007f984221ce21 in run_trans_callbacks (type=type@entry=8, trans=trans@entry=0x7f9842c6ae38, req=<optimized out>, rpl=rpl@entry=0x7f98448ff238, code=code@entry=180) at t_hooks.c:209
        params = {req = 0x7f9842c6c8d8, rpl = 0x7f98448ff238, code = 180, param = 0x7f9842c819d8, extra1 = 0x0, extra2 = 0x0}
        cbp = 0x7f9842c819c8
        backup = 0x7e4860 <global_avps>
        trans_backup = 0x7f9842c6ae38
        __FUNCTION__ = "run_trans_callbacks"
#7  0x00007f984222aa3a in relay_reply (t=0x7f9842c6ae38, p_msg=<optimized out>, branch=0, msg_status=180, cancel_bitmap=0x7fff55e57100) at t_reply.c:1173
        relay = 0
        save_clone = 0
        buf = 0x0
        res_len = 0
        relayed_code = 0
        relayed_msg = 0x7f98448ff238
        bm = {to_tag_val = {s = 0x1 <error: Cannot access memory at address 0x1>, len = 4345328}}
        totag_retr = 0
        reply_status = RPS_PROVISIONAL
        uas_rb = 0x7f9842c6af38
        cb_s = {s = 0x195 <error: Cannot access memory at address 0x195>, len = 0}
        text = {s = 0x7f98448ff238 "\005", len = 0}
        __FUNCTION__ = "relay_reply"
#8  0x00007f984222e545 in reply_received (p_msg=0x7f98448ff238) at t_reply.c:1505
        msg_status = 180
        last_uac_status = 100
        branch = 0
        reply_status = <optimized out>
        timer = 140291962040888
        cancel_bitmap = 0
        uac = 0x7f9842c6b010
        t = 0x7f9842c6ae38
---Type <return> to continue, or q <return> to quit---
        backup_list = <optimized out>
        has_reply_route = <optimized out>
        __FUNCTION__ = "reply_received"
#9  0x000000000043b60c in forward_reply (msg=msg@entry=0x7f98448ff238) at forward.c:495
        new_buf = 0x0
        to = 0x0
        new_len = 0
        mod = 0x7f98448ac770
        proto = <optimized out>
        id = 0
        send_sock = <optimized out>
        s = <optimized out>
        len = <optimized out>
        __FUNCTION__ = "forward_reply"
#10 0x00000000004720b2 in receive_msg (
    buf=0x7f983f583da8 <wss_current_req+8> "SIP/2.0 180 Ringing\r\nVia: SIP/2.0/WSS 91.236.192.35:5066;branch=z9hG4bKfeac.6e5b49a4.0\r\nTo: <sip:31103027540@89.35.18.21:5065>;tag=74c0band8s\r\nFrom: \"0031614720622\" <sip:0031614720622@motto.nl:5080>;t"..., len=<optimized out>, len@entry=405, rcv_info=rcv_info@entry=0x7fff55e572e0, existing_context=existing_context@entry=0x0) at receive.c:257
        ctx = 0x7f98448ffdd8
        msg = 0x7f98448ff238
        start = {tv_sec = 1491806248, tv_usec = 446592}
        rc = 3
        in_buff = {
          s = 0x7f983f583da8 <wss_current_req+8> "SIP/2.0 180 Ringing\r\nVia: SIP/2.0/WSS 91.236.192.35:5066;branch=z9hG4bKfeac.6e5b49a4.0\r\nTo: <sip:31103027540@89.35.18.21:5065>;tag=74c0band8s\r\nFrom: \"0031614720622\" <sip:0031614720622@motto.nl:5080>;t"..., len = 405}
        __FUNCTION__ = "receive_msg"
#11 0x00007f983f37b8d0 in ws_process (con=0x7f9842c69aa8) at ../proto_ws/ws_common.h:533
        size = 0
        req = 0x7f983f583da0 <wss_current_req>
        ret_code = WS_ERR_NONE
        bk = 51 '3'
        msg_len = 405
        local_rcv = {src_ip = {af = 2, len = 4, u = {addrl = {492837204, 0}, addr32 = {492837204, 0, 0, 0}, addr16 = {6484, 7520, 0, 0, 0, 0, 0, 0}, addr = "T\031`\035", '\000' <repeats 11 times>}}, dst_ip = {
            af = 2, len = 4, u = {addrl = {599845979, 0}, addr32 = {599845979, 0, 0, 0}, addr16 = {60507, 9152, 0, 0, 0, 0, 0, 0}, addr = "[\354\300#", '\000' <repeats 11 times>}}, src_port = 18038, 
          dst_port = 5066, proto = 6, proto_reserved1 = 200130223, proto_reserved2 = 0, src_su = {s = {sa_family = 2, sa_data = "FvT\031`\035\000\000\000\000\000\000\000"}, sin = {sin_family = 2, sin_port = 30278, 
              sin_addr = {s_addr = 492837204}, sin_zero = "\000\000\000\000\000\000\000"}, sin6 = {sin6_family = 2, sin6_port = 30278, sin6_flowinfo = 492837204, sin6_addr = {__in6_u = {
                  __u6_addr8 = "\000\000\000\000\000\000\000\000\020`\252\000\000\000\000", __u6_addr16 = {0, 0, 0, 0, 24592, 170, 0, 0}, __u6_addr32 = {0, 0, 11165712, 0}}}, sin6_scope_id = 1160185045}}, 
          bind_address = 0x7f98448b36f0}
        newreq = <optimized out>
        msg_buf = <optimized out>
#12 wss_read_req (con=0x7f9842c69aa8, bytes_read=<optimized out>) at proto_wss.c:582
        size = <optimized out>
        d = <optimized out>
---Type <return> to continue, or q <return> to quit---
        __FUNCTION__ = "wss_read_req"
#13 0x00000000005392c9 in handle_io (fm=<optimized out>, idx=idx@entry=0, event_type=event_type@entry=1) at net/net_tcp_proc.c:227
        ret = 0
        n = <optimized out>
        con = 0x7f9842c69aa8
        s = 6
        rw = <optimized out>
        resp = <optimized out>
        response = {140291932068520, 1}
        __FUNCTION__ = "handle_io"
#14 0x000000000053c7df in io_wait_loop_epoll (h=<optimized out>, t=<optimized out>, repeat=<optimized out>) at net/../io_wait_loop.h:225
        ret = <optimized out>
        e = <optimized out>
        n = 1
        r = 0
#15 tcp_worker_proc_loop () at net/net_tcp_proc.c:355
        __FUNCTION__ = "tcp_worker_proc_loop"
#16 0x0000000000534966 in tcp_start_processes (chd_rank=chd_rank@entry=0x7c8d40 <chd_rank>, startup_done=startup_done@entry=0x0) at net/net_tcp.c:1831
        r = 0
        reader_fd = {47, 50}
        pid = <optimized out>
        load_p = 0x7f9842bb4df0
        __FUNCTION__ = "tcp_start_processes"
#17 0x000000000041c3f3 in main_loop () at main.c:683
        startup_done = 0x0
        chd_rank = 9
#18 main (argc=<optimized out>, argv=<optimized out>) at main.c:1283
        cfg_stream = <optimized out>
        c = <optimized out>
        r = <optimized out>
        tmp = 0x1 <error: Cannot access memory at address 0x1>
        tmp_len = <optimized out>
        port = <optimized out>
        proto = <optimized out>
        protos_no = <optimized out>
        options = 0x567c78 "f:cCm:M:b:l:n:N:rRvdDFETSVhw:t:u:g:P:G:W:o:"
        ret = -1
        seed = 884085485
        __FUNCTION__ = "main"
```

**Comment by razvancrainea** [comment]
Just figured out where the crash comes from. I am pushing a fix right away.

**Comment by razvancrainea** [comment]
Closed by 9e251ce
